windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -753,18 +755,29 @@ procedure DownloadingState.Enter;
 var
   DownloadResult: Boolean;
   RetryCount: Integer;
+  FMutex: TKeymanMutex;
 begin
   // Enter DownloadingState
   bucStateContext.SetRegistryState(usDownloading);
 
-  RetryCount := 0;
-  DownloadResult := False;
+  FMutex := TKeymanMutex.Create(KeymanDownloadMutexName);
 
-  while (not DownloadResult) and (RetryCount < 3) do
-  begin
-    DownloadResult := DownloadUpdatesBackground;
-    if not DownloadResult then
+  try
+    // Should be impossible but just exit anyway and let the process current
+    // downloading process finish.
+    if not FMutex.TakeOwnership then
+    begin
+      Exit;
+    end;
+
+    RetryCount := 0;
+    repeat
+      DownloadResult := DownloadUpdatesBackground;
       Inc(RetryCount);
+    until DownloadResult or (RetryCount = 3);
```

Comment:
I'd make it `(RetryCount >= 3)` which would be clearer even though it can't happen.

```suggestion
    until DownloadResult or (RetryCount >= 3);
```